proposals/NNNN-Continuous-Integration-And-Testing.md

```diff
@@ -0,0 +1,190 @@
+| **NIGHTLY**<br>Build check | **WEEKLY**<br>Build check | **PullRequest**<br>Build check | **Push**<br>Build check |
+|:--------:|:-------------:|:-----------:|:-----------:|
+|   BUILD_TESTS=OFF   |               |             |   BUILD_TESTS=OFF   |
+|    BUILD_TESTS=ON    |               |     BUILD_TESTS=ON      |    BUILD_TESTS=ON    |
+|          | BUILD_TESTS=ON<br>ENABLE_GCOV=ON<br>+ Coverage |             |          |
+
+CI ATF jobs:
+
+|**ATF TEST SET**|**Trigger**|
+|:----------------------:|:--- |
+|Smoke|noUT,<br>UT (for PullRequest jobs only)|
+|Feature|noUT|
+|Regression|noUT|
+
+- **Smoke** - Run ATF tests from the Smoke test set (basic project checks and happy paths of all main features)
+- **Feature** - Run ATF tests from the Feature test set (such test set is created for each feature, and includes tests for checking functionality of a new feature). It is applicable only for Feature branches.
+- **Regression** - Run ATF tests from all Feature test sets, which are already merged to the develop branch.
+
+### 1. Optimizing of building process and resources usage
+
+To optimize the CI processes and resources usage it is necessary to separate the third-party libraries building from the SDL building. Further, the third-party libraries as preloaded files should be used for all testing related tasks. This approach saves server time and can allow getting the CI status faster.
+
+### 2. Using cluster for a High Availability and Fault Tolerance of the CI processes
+
+To minimize risks of fails, and to optimize CI processes we need to use the Best Practices of computation, and keep project infrastructure on the top level using **Mesos** cluster and **Docker** container-based services. Containers allow building and testing applications in standard, portable ways.
+A server cluster is a centralized fault-tolerant system for managing resources. The use of the cluster is appropriate for distributed computing environments in order to ensure the isolation of resources and convenient management. The cluster allows allocating CPU, memory, and other resources for any tasks like CI jobs.
+
+For example **Mesos** can be used as a cluster core to allocate an environment for working tasks. Such frameworks as: Jenkins, Spark, Marathon, Aurora can be used to implement the logic of launching tasks, monitoring their work, scaling, etc.
+
+When the product is growing up, and count of the same time features is growing up too, wee need to have an ability to **scale our CI system fast and without any downtime**. The key to success here is usage of cluster.
+
+Another one point is a **Foult Tolerance**. When you're using a cluster systems you are completely safe from disk failures or even entire server crashes.
+
+Schema of cluster
+![img][cluster_schema]
+
+## Potential downsides
+
+No obvious places can be affected.
+
+## Impact on existing code
+
+Can be expected a lot of places in scope of fixing existing issues reported by static/dynamic analysis tools.
```

Comment:
```suggestion
It can be expected that there will be a lot of places in scope of fixing existing issues reported by static/dynamic analysis tools.
```